### PR TITLE
feat: add default branch prop for publish:github action

### PR DESCRIPTION
**Pull request by shoukoo**
Closes https://github.com/backstage/backstage/issues/6310

## Hey, I just made a Pull Request!
It has become common practice to use main as the default branch when initializing a Git repository. This updates the Publish:GitHub action to have the defaultBranch config property. If not provided, master will still be used.

The user will be able to change the default branch in your template like below.
```
- id: publish
   name: Publish
   action: publish:github
   input:
     allowedHosts: ['github.com']
     description: 'This is {{ parameters.name }}'
     repoUrl: '{{ parameters.repoUrl }}'
     defaultBranch: 'main' << Add this to change your default branch.
```

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-common/src/scm/git.ts [hunk 1]
```diff
@@ -150,12 +150,19 @@ export class Git {
     });
   }
 
-  async init({ dir }: { dir: string }): Promise<void> {
+  async init({
+    dir,
+    defaultBranch,
```
Comment (benjdlambert) [inline packages/backend-common/src/scm/git.ts:155]: I wonder if we should set a `defaultBranch = something` here too in order to not break this public API?

Comment (shoukoo) [reply packages/backend-common/src/scm/git.ts]: Make sense! do you also want me to add the defaultBranch property in other providers as well?

Comment (freben) [reply packages/backend-common/src/scm/git.ts]: That makes sense, and yeah separate pr sounds good

File: packages/backend-common/src/scm/git.ts [hunk 2]
```diff
@@ -150,12 +150,19 @@ export class Git {
     });
   }
 
-  async init({ dir }: { dir: string }): Promise<void> {
+  async init({
+    dir,
+    defaultBranch = 'master',
+  }: {
+    dir: string;
+    defaultBranch: string;
```
Comment (benjdlambert) [inline packages/backend-common/src/scm/git.ts:158]: I /think/ that this should be `defaultBranch?: string` so then typescript knows that you don't have to specify it. The API report still seems to suggest that it's required when it should not be :)

Comment (shoukoo) [reply packages/backend-common/src/scm/git.ts]: Thanks @benjdlambert , just updated the code.

File: .changeset/heavy-colts-approve.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/backend-common': minor
```
Comment (benjdlambert) [inline .changeset/heavy-colts-approve.md:2]: think that these are still `patch` changes though not `minor` since we are `<1.0.0`

Comment (shoukoo) [reply .changeset/heavy-colts-approve.md]: of course! updated.

File: packages/backend-common/src/scm/git.ts (hunk 2 above)
Comment (cal5barton) [reply packages/backend-common/src/scm/git.ts]: Where would this be passed, from the UI or can this be set in the app configuration?

Comment (shoukoo) [reply packages/backend-common/src/scm/git.ts]: Hey @cal5barton, you can pass in the default branch to the publish:github action. so you will be able to do something like this in your template to change the default branch:

```
- id: publish
      name: Publish
      action: publish:github
      input:
        allowedHosts: ['github.com']
        description: 'This is {{ parameters.name }}'
        repoUrl: '{{ parameters.repoUrl }}'
        defaultBranch: 'main' << Add this to change your default branch.
```

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**